[WB-43] Added caching with redis

charlieduong94 opened the pull request:
Tweaked tests to run within docker container
Modified .travis.yml to run tests in docker

charlieduong94 commented on test/integration/queue/producers/QueueProducer-test.js:
```diff
@@ -14,7 +14,7 @@ const messageParser = require('~/queue/util/message-parser')
 const waitForEvent = require('~/test/util/waitForEvent')
 
-const AMQ_URL = 'amqp://localhost'
+const AMQ_URL = 'amqp://rabbitmq'
```
These should all be moved to some sort of test config. I'll handle this in a future PR since this is already getting too large.

Druotic replied:
Hmm, it works with either `amqp:rabbitmq` or `amqp://rabbitmq`? Surprising, but a future pr that consolidates it would be nice.

charlieduong94 commented on Dockerfile:
```diff
@@ -0,0 +1,16 @@
+RUN cd /tmp && npm install
+
+RUN mkdir $HOME/node_modules
+RUN cp -a /tmp/node_modules $HOME
```
This part is cool little trick. We add the `package.json` to a temp dir and install modules in that dir. Later on we copy the files over our working directory. The next time the build runs, docker will diff the temp dir against what was cached to determine if a fresh install is needed. The only time a change will be detected is if the `package.json` is modified.

charlieduong94 commented on package.json:
```diff
@@ -3,14 +3,15 @@
+    "docker:build": "docker-compose build",
+    "docker:start": "docker-compose up rediscluster rabbitmq",
     "docker:stop": "docker-compose down",
-    "pretest": "npm run lint",
```
I removed this `pretest` hook because I wanted to allow for all dev to happen within docker containers.

charlieduong94 replied:
The linting will still be run when the regular test runs.

charlieduong94 commented on docker-compose.yml:
```diff
@@ -1,6 +1,21 @@
+      - rabbitmq
+    volumes:
+      - .:/usr/windbreaker
+      - /usr/windbreaker/node_modules
```
Another neat trick. We mount the the service utils into the container. Specifying the `node_modules` folder like this forces docker to remount whatever was originally in that location (the modules installed from the build). This let's us make quick changes without having to do rebuilds.

charlieduong94 replied:
For a faster feedback loop, I recommend running `docker-compose run test bash` to get access to container itself. You can make changes from the host and run tests from within the container.

charlieduong94 replied:
I'll add docs for that in the morning.

charlieduong94 commented on cache/util/getRetryStrategy.js:
```diff
@@ -0,0 +1,12 @@
+  return function (retryCount) {
+    return Math.min(1000 + (retryCount * 1000), maximumDelay)
+  }
+}
```
I opted to use a simple retry strategy instead of managing the reconnections manually. If we find the need to manually handle the reconnects, we can switch it out later.

Druotic commented on cache/util/getRetryStrategy.js:
```diff
@@ -0,0 +1,12 @@
+// simple retry strategy
+
+// default max delay in milliseconds
+const DEFAULT_MAX_DELAY = 10000
```
Do we really want to give up after 10 secs? Seems rather short.

charlieduong94 replied:
Not really giving up after 10 secs, but at the longest period, retrying after 10 secs. I suppose that's still short though, I can bump the max up.

Druotic commented on cache/util/getRetryStrategy.js:
```diff
@@ -0,0 +1,12 @@
+  const maximumDelay = maxDelay || DEFAULT_MAX_DELAY
+
+  return function (retryCount) {
+    return Math.min(1000 + (retryCount * 1000), maximumDelay)
```
I know you were trying to keep this simple for now, but maybe add an element of randomness (jitter). Parallel retries will all happen at the same time if the delays are all the same.

Druotic commented on docker-compose.yml:
```diff
@@ -1,6 +1,21 @@
   rabbitmq:
     image: 'rabbitmq:3.6.10-alpine'
     ports:
       - '5672:5672'
```
Do these ports still need to be exposed, or is this just for debugging purposes?

charlieduong94 replied:
Just for debugging purposes.

Druotic commented on cache/Cache.js:
```diff
@@ -0,0 +1,50 @@
+
+    redisClient.on('error', (err) => {
+      this.emit('error', err)
+    })
```
It might be worth listening/logging other events - e.g. 'node error', '+node', '-node', 'reconnecting'.

charlieduong94 replied:
I'll add listeners for `reconnecting` and `node error`. I don't feel like clients need to know when nodes are added or removed though.

Druotic commented on cache/Cache.js:
```diff
@@ -0,0 +1,50 @@
+const _createRedisClient = require('./util/createRedisClient')
+const msgpack = require('msgpack-lite')
+
+const DEFAULT_EXPIRATION = 86400
```
Units?

charlieduong94 replied:
seconds, will add comment

Druotic commented on .travis.yml:
```diff
@@ -1,13 +1,7 @@
-  - rabbitmq
 
-script: npm run test:ci
+script: npm run test
```
Should be able to just use `npm test` here. Also - are we guaranteed `npm` is available if we don't specify the language/node_js version? For consistent behavior, might still be worth specifying.

charlieduong94 replied:
I'll switch back to specifying the language just to be safe.

charlieduong94 replied:
better yet, let's just use `docker-compose run test` here. That will help us avoid the initial node modules install that travis automatically performs.

austinkelleher commented on cache/Cache.js:
```diff
@@ -0,0 +1,63 @@
+    this._onReady = () => this.emit('ready')
+    this._onError = (err) => this.emit('error', err)
+    this._onReconnecting = () => this.emit('reconnecting')
+    this._onNodeError = (err) => this.emit('node error', err)
```
I'm personally not a big fan of events with spaces. Maybe `node-error`? What do you think?

charlieduong94 replied:
I could add the dash. I was just trying to match up with what `ioredis` emits.

charlieduong94 replied:
I added the dash since it matches up with our other event emitters.

austinkelleher commented on cache/Cache.js:
```diff
@@ -0,0 +1,63 @@
+class Cache extends EventEmitter {
+  constructor ({ logger, nodes, defaultTtl, redisClientOptions }) {
+    super()
+    this._logger = logger
```
This is currently unused. It should probably also be passed to `conflogger.configure(...)`

charlieduong94 replied:
Fixed

austinkelleher commented on cache/Cache.js:
```diff
@@ -0,0 +1,63 @@
+  }
+
+  async get (key) {
+    const encoded = await this._redisClient.getBuffer(key)
```
What do you think about adding a `try/catch` around these `await this._redisClient.*` functions and log an error + throw it afterward? Then we wouldn't have to add log errors in all of the services. The logger that is passed could eventually have appenders for different log services (Sumo etc.) or write to file automatically.

```js
async get (key) {
  let encoded
  try {
    encoded = this._redisClient.getBuffer(key)
  } catch (err) {
    // logger.error could have an error appender for Sumo, write to file, etc. We wouldn't have to log out the error
    // everywhere that we use cache service calls
    this._logger.error(`Error occurred fetching key "${key}"`, err)
    throw err
  }
  return encoded ? msgpack.decode(encoded) : encoded
}
```

charlieduong94 replied:
Great suggestion. The additional logging is would be helpful.

charlieduong94 replied:
Fixed.

austinkelleher commented on cache/Cache.js:
```diff
@@ -0,0 +1,63 @@
+
+  async getEntity (key, ModelType) {
+    const data = await this.get(key)
+    return data ? ModelType.wrap(data) : data
```
Should never happen, but technically if data is corrupted, `ModelType.wrap()` could have errors. We should add an array of errors as the second arg and log/throw if `errors.length`:

```js
async getEntity (key, ModelType) {
  const data = await this.get(key)
  if (!data) return data
  
  let errors = []
  data = ModelType.wrap(data, errors)
  
  if (errors.length) {
    logger.error(`Errors occurred while wrapping model: "${errors.join(',')}"`)
    return // If data is corrupt, we should return undefined
  }
  return data
}
```

Druotic replied:
Returning undefined on corrupted data seems dangerous. Why not throw an error?

austinkelleher replied:
Yes I agree.

charlieduong94 replied:
Will fix. Thanks

charlieduong94 replied:
fixed

austinkelleher commented on cache/Cache.js:
```diff
@@ -0,0 +1,63 @@
+      .removeListener('node error', this._onNodeError)
+      .removeListener('error', this._onError)
+
+    this._redisClient.disconnect()
```
Not sure if this is what we want, but `redis.quit()` waits for pending replies to finish before closing. It's more graceful, but if we go that way we need to make sure that we don't infinitely wait for replies that never get here.

charlieduong94 replied:
I switched to `quit`. I'm not sure how to ensure that we don't indefinitely wait aside from adding a timeout. If you would like, I could add the timeout, but for now I think this is sufficient.

austinkelleher commented on cache/util/createCache.js:
```diff
@@ -0,0 +1,21 @@
+const Cache = require('../Cache')
+
+module.exports = async function createCache (nodes, options) {
+  const cache = new Cache(nodes, options)
```
Maybe I'm missing something, but your method signature in `Cache` suggests that there is only one parameter. 

```js
class Cache extends EventEmitter {
  constructor ({ logger, nodes, defaultTtl, redisClientOptions }) {}
}
```

charlieduong94 replied:
Thanks for catching this. Will fix.

charlieduong94 replied:
Fixed

charlieduong94 commented:
@Druotic @austinkelleher Thanks for reviewing, I addressed the feedback I was given. If there is anything that you feel needs to be done, let me know.

charlieduong94 commented:
I rebased against master to pull in @austinkelleher 's latest changes. Please give it a once over to make sure I haven't missed anything.

charlieduong94 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Consider adding caching layer with redis](https://trello.com/c/rHMZrKen/43-consider-adding-caching-layer-with-redis)
